### PR TITLE
PromQL: Fix bug with inconsistent results for queries with OR expression and EnableDelayedNameRemoval

### PR DESCRIPTION
**Pull request by zenador**
There's a bug with inconsistent results with OR expressions when the [experimental feature `EnableDelayedNameRemoval`](https://github.com/prometheus/prometheus/pull/14477) is on.

With the added flaky tests in this PR, if you run the test once, it generally passes, but if you run it repeatedly, it sometimes fails.

Try it with `go test -v ./promql -run "TestEvaluations/testdata/name_label_dropping_2/line_" --count=100`

The order of series processed is not deterministic, and depending on the order, `dropName` may or may not be applied, so with the same query and data, sometimes the metric name is dropped and sometimes it's not.

This is because we use a map to produce the results in the output matrix for the binary operation, and native maps in golang do not enforce order when ranging over them, and this matrix is passed to the sum aggregation where the order of input determines the dropName applied for the grouping.

This PR does introduce some small overhead in an extra slice to store the hashes in the original order. Not sure if this is small enough to be worth it to fix the inconsistent results or if we should find a better solution.

#### Does this PR introduce a user-facing change?

```release-notes
[BUGFIX] PromQL: Fix bug with inconsistent results for queries with OR expression and the experimental EnableDelayedNameRemoval. #17161
```

### CODE REVIEW

File: promql/promqltest/testdata/name_label_dropping_2.test [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Test for __name__ label drop.
+load 1m
+  metric_total{env="1"} 0+1x10
+  metric_total{env="2"} 0+3x10
+
+eval instant at 10m sum by (__name__) (metric_total{env="1"} or rate(metric_total{env="2"}[5m]))
+  metric_total 10.05
+
+eval instant at 10m sum by (__name__) (rate(metric_total{env="2"}[5m]) or metric_total{env="1"})
+  {} 10.05
```
Comment (charleskorn) [inline promql/promqltest/testdata/name_label_dropping_2.test]: Should the result series be `metric_total`, rather than `{}`? It seems weird that the grouping label from the `sum` is dropped.

Comment (zenador) [reply promql/promqltest/testdata/name_label_dropping_2.test]: I think this part is fine, current Prometheus behaviour is pretty consistent with dropping this, preserving it is only allowed with a label function, e.g. the following passes consistently even 1000 times:

```
# Test for __name__ label drop.
load 1m
  metric_total{env="1"} 0+1x10
  metric_total{env="2"} 0+3x10

# Metric name is preserved as there is no function that drops it.
eval instant at 10m sum by (__name__) (metric_total{env="1"})
  metric_total 10

# Metric name is dropped at the end because of rate and because there is no label function to preserve it.
eval instant at 10m sum by (__name__) (rate(metric_total{env="2"}[5m]))
  {} 0.05

# Metric name is preserved with label_replace even though it would have been dropped with rate.
eval instant at 10m label_replace(sum by (__name__) (rate(metric_total{env="2"}[5m])), "__name__", "$1", "__name__", "(.+)")
  metric_total 0.05
```

File: promql/promqltest/testdata/name_label_dropping_2.test [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# SPDX-License-Identifier: AGPL-3.0-only
```
Comment (roidelapluie) [inline promql/promqltest/testdata/name_label_dropping_2.test]: This is going to be an issue, as the codebase is Apache 2.

Comment (zenador) [reply promql/promqltest/testdata/name_label_dropping_2.test]: Good catch, thanks! Sorry about that, didn't pay enough attention to licensing details.

**Comment by beorn7** [comment]
Thanks for catching this.

Formal nit: Let's not create `name_label_dropping_2.test` but just append to the existing `name_label_dropping.test`.

About the real meat here: This is an interesting edge case you have discovered here (although mostly an academic one, as you have to aggregate by `__name__` (not very common) and have to construct a caso where a name marked for dropping is equal to a name not marked for dropping (which I would expect to happen only in contrived examples or by accident rather than in a query that actually serves any purpose)). Nevertheless, we need to handle this case in a defined and reproducible way.

The approach here is reproducible, but I would say it is defined in a problematic way: IIUC, the decision whether the name is ultimately dropped or not is made by the series that happens to land on top of the list. Which is hard to predict by the user, if at all. It also doesn't seem to make a lot of sense (although there is maybe not much need to make a lot of sense for this rare edge case). Still, this worries me much more than the performance concerns you have already brought up.

The core of the problem here is that we aggregate (by name) a number of series with the same name where some of the names are marked as to be dropped and others are not. The fundamental question is what should we do with the name, drop or not? Which gives us two obvious ways, and then I also came up with a trickier way out:

1. In that case, we always drop the name.
2. In that case, we always keep the name.
3. We somehow consider a name marked for dropping to have a different identity than a name not marked for dropping, so we aggregate the series into two different groups after all, and since one of the groups gets its name dropped in the end, we don't even produce an ambiguity.

I believe (1) and (2) are easy to implement and fairly clear. I cannot honestly say which one I like better. It's hard to argue for one over the other, given that there is no real-world use case to judge them.

I like (3) most because it makes sense semantically and is close to the behavior we have without the delayed name dropping. I'm just concerned that it will have deeper implications and might not be trivial to implement.

What do others think? @jcreixell @juliusv @roidelapluie might be the people with the most context, but I'd be happy about anyone chiming in with insights.

**Comment by zenador** [comment]
> Formal nit: Let's not create name_label_dropping_2.test but just append to the existing name_label_dropping.test.

Updated.

> the decision whether the name is ultimately dropped or not is made by the series that happens to land on top of the list. Which is hard to predict by the user, if at all.

True, added some test cases to demonstrate this further. If the LHS of the OR expression has no series, we will use RHS to determine the name dropping, but otherwise we will use LHS. But this depends on the underlying data and not the input expression, and it seems strange to have the labels change unpredictably based on that.

Regarding your suggestions, personally I'd vote against 3 as it is added complexity for something which is currently just academic (unless someone thinks of a real-world use case where implementing 3 would actually be preferable).

I like either 1 or 2, no preference as long as we apply it consistently. Just to be clear, when we have multiple series, we'd continue to drop the name if all series drop the name, and continue to not drop the name if all series not drop the name. But when not all the series behave the same way:

1) If at least 1 series drops the name, we drop the name for all series.
2) If at least 1 series keeps the name, we keep the name for all series

But actually, we'd run into the same problem above, where the resulting labels depend on the underlying data and is not predicable from the input expression alone, as for instance in `sum by (__name__) (metric_total{env="3"} or rate(metric_total{env="2"}[5m]))`, we wouldn't know without running the query whether we'd get series from the LHS or RHS or both or neither. So it looks like we can't solve this problem without the additional complexity of 3, but at least choosing 1 or 2 over the current approach in this PR saves us from having to use extra memory and makes it independent of the order.

**Comment by beorn7** [comment]
I broadly agree with your conclusion that we should pick the least complicated solution, as there will be hardly any real-world use cases that would depend on a particular behavior being picked over another.

I'm still not so sure how problematic approach (3) would be in this regard, but let's assume it would be the most complicated ones, with possibly some surprising implications further down the road.

If the choice is between (1) and (2), I'd currently prefer (1), i.e. "If at least 1 series drops the name, we drop the name for all series.", following the rationale that with at least one thing in the aggregation that doesn't want to be called by its old name anymore, the aggregate of everything probably doesn't like that name either. (Or: If I add one apple to a lot of oranges, I cannot claim anymore that the aggregate is still oranges.)

So let's go for (1).

**Comment by beorn7** [comment]
> But actually, we'd run into the same problem above, where the resulting labels depend on the underlying data and is not predicable from the input expression alone, as for instance in `sum by (__name__) (metric_total{env="3"} or rate(metric_total{env="2"}[5m]))`, we wouldn't know without running the query whether we'd get series from the LHS or RHS or both or neither.

I don't quite understand this concern. Don't we decide about the name dropping dynamically anyway, i.e. while evaluating the expression, rather than ahead of the evaluation just based on syntax?

**Comment by beorn7** [comment]
About properly reviewing this: I might only get to it after PromCon. But if I understand correctly, you would first implement approach (1) before it makes sense to do a detailed review anyway.

**Comment by beorn7** [comment]
BTW: This behavior should be included in the feature flag documentation.

File: docs/feature_flags.md [hunk 3]
```diff
@@ -143,6 +143,9 @@ These queries are rare to occur and easy to fix. (In the above example,
 removing `by (__name__)` doesn't change anything without the feature flag and
 fixes the possible problem with the feature flag.)
 
+When we have multiple series for the same grouping key that do not agree on the same dropping
+name behaviour, if at least 1 series drops the name, we drop the name for all series.
```
Comment (beorn7) [inline docs/feature_flags.md]: I think we need to mention that this happens in aggregations. And that this is a rare case so that people don't think too much about it. Suggestion:

```suggestion
It is possible to craft a query that aggregates by `__name__` and puts samples with and without delayed name removal into the same group. In that case, the name is removed from the affected group. Note that this case hardly occurs in queries that fulfill a practical purpose.
```

File: promql/engine.go [hunk 4]
```diff
@@ -3172,6 +3172,7 @@ type groupedAggregation struct {
 	incrementalMean        bool // True after reverting to incremental calculation of the mean value.
 	counterResetSeen       bool // Counter reset hint CounterReset seen. Currently only used for histogram samples.
 	notCounterResetSeen    bool // Counter reset hint NotCounterReset seen. Currently only used for histogram samples.
+	dropName               bool // True if any series in this group have DropName set.
```
Comment (beorn7) [inline promql/engine.go]:
```suggestion
	dropName               bool // True if any sample in this group has DropName set.
```

For consistency. Also, grammar nit fixed.

**Comment by zenador** [comment]
Thank you and no worries about the delay! Applied your suggestions.

Realised I forgot to reply this:

> I don't quite understand this concern. Don't we decide about the name dropping dynamically anyway, i.e. while evaluating the expression, rather than ahead of the evaluation just based on syntax?

Yes that's true for this edge case. But typically for normal cases we would be able to tell just by looking, e.g. `metric_total{env="3"}` does not drop name but `rate(metric_total{env="2"}[5m])` does. I don't have a better suggestion though.

**Comment by zenador** [comment]
I believe the windows test failure is transient (or at least it should have nothing to do with the changes in this PR).

**Comment by beorn7** [comment]
Thank you very much. Indeed, the MS Windows tests are often flaky. I'll re-run, but if it doesn't work, I'll just override.
